File: trading_system/runtime_config.py
from __future__ import annotations

import json
import os
from pathlib import Path

REPO_ROOT = Path(os.environ.get('PERPS_BOT_REPO_ROOT', Path(__file__).resolve().parents[1]))
CONFIG_FILE = Path(os.environ.get('PERPS_BOT_CONFIG_FILE', str(REPO_ROOT / 'config' / 'local.json')))

try:
    _raw = json.loads(CONFIG_FILE.read_text()) if CONFIG_FILE.exists() else {}
    CONFIG = _raw if isinstance(_raw, dict) else {}
except Exception:
    CONFIG = {}


def cfg(key: str, fallback: str) -> str:
    return os.environ.get(key) or CONFIG.get(key) or fallback

BOT_HOME = Path(cfg('HERMES_PERPS_BOT_HOME', os.environ.get('HOME', str(REPO_ROOT))))
DATA_DIR = Path(cfg('PERPS_AUTO_TRADE_DATA_DIR', str(BOT_HOME / '.trading-data')))
DB_PATH = Path(cfg('PERPS_AUTO_TRADE_DB_PATH', str(DATA_DIR / 'trading.db')))
BRIDGE_DIR = DATA_DIR / 'telegram-bridge'
TELEGRAM_TOKEN_FILE = Path(cfg('TELEGRAM_TOKEN_FILE', str(REPO_ROOT / 'telegram.txt')))
TELEGRAM_TRADE_CHAT_ID = cfg('TELEGRAM_TRADE_CHAT_ID', '123456789')
JUP_BIN = cfg('JUP_BIN', 'jup')
HELIUS_BIN = cfg('HELIUS_BIN', 'helius')
PERPS_MONITOR_PATH = Path(cfg('PERPS_MONITOR_PATH', str(REPO_ROOT / 'perps-monitor.mjs')))
PERPS_EXECUTOR_PATH = Path(cfg('PERPS_EXECUTOR_PATH', str(REPO_ROOT / 'perps-auto-trade.mjs')))
PERPS_RUNTIME_ENV_FILE = Path(cfg('PERPS_RUNTIME_ENV_FILE', str(DATA_DIR / 'perps-runtime.env')))
PERPS_WALLET_ADDRESS = cfg('PERPS_WALLET_ADDRESS', 'REPLACE_ME_WALLET_ADDRESS')
SESSIONS_INDEX = REPO_ROOT / '.hermes' / 'sessions' / 'sessions.json'
SESSIONS_DIR = REPO_ROOT / '.hermes' / 'sessions'


def build_path_env() -> str:
    extras = []
    home = os.environ.get('HOME')
    if os.environ.get('JUP_BIN_DIR') or CONFIG.get('JUP_BIN_DIR'):
        extras.append(os.environ.get('JUP_BIN_DIR') or CONFIG.get('JUP_BIN_DIR'))
    if os.environ.get('HELIUS_BIN_DIR') or CONFIG.get('HELIUS_BIN_DIR'):
        extras.append(os.environ.get('HELIUS_BIN_DIR') or CONFIG.get('HELIUS_BIN_DIR'))
    if home:
        extras.append(str(Path(home) / '.hermes' / 'node' / 'bin'))
        extras.append(str(Path(home) / '.cargo' / 'bin'))
    extras = [p for p in extras if p]
    return ':'.join(extras + [os.environ.get('PATH', '')]) if extras else os.environ.get('PATH', '')

File: runtime-config.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
export const REPO_ROOT = path.dirname(__filename);
const CONFIG_FILE = process.env.PERPS_BOT_CONFIG_FILE || path.join(REPO_ROOT, 'config', 'local.json');

function readConfig() {
  try {
    if (!fs.existsSync(CONFIG_FILE)) return {};
    const raw = JSON.parse(fs.readFileSync(CONFIG_FILE, 'utf-8'));
    return raw && typeof raw === 'object' ? raw : {};
  } catch {
    return {};
  }
}

const CONFIG = readConfig();
const cfg = (key, fallback) => process.env[key] || CONFIG[key] || fallback;

export const BOT_HOME = cfg('HERMES_PERPS_BOT_HOME', process.env.HOME || REPO_ROOT);
export const DATA_DIR = cfg('PERPS_AUTO_TRADE_DATA_DIR', path.join(BOT_HOME, '.trading-data'));
export const DB_PATH = cfg('PERPS_AUTO_TRADE_DB_PATH', path.join(DATA_DIR, 'trading.db'));
export const DB_CLI = cfg('PERPS_AUTO_TRADE_DB_CLI', path.join(REPO_ROOT, 'trading_system', 'perps_db_cli.py'));
export const JUP_BIN = cfg('JUP_BIN', 'jup');
export const HELIUS_BIN = cfg('HELIUS_BIN', 'helius');
export const TELEGRAM_TOKEN_FILE = cfg('TELEGRAM_TOKEN_FILE', path.join(REPO_ROOT, 'telegram.txt'));
export const TELEGRAM_TRADE_CHAT_ID = cfg('TELEGRAM_TRADE_CHAT_ID', '123456789');
export const DEFAULT_WALLET_ADDRESS = cfg('PERPS_WALLET_ADDRESS', 'REPLACE_ME_WALLET_ADDRESS');
export const PERPS_RUNTIME_ENV_FILE = cfg('PERPS_RUNTIME_ENV_FILE', path.join(DATA_DIR, 'perps-runtime.env'));
export const TRADING_PLAYBOOK_PATH = cfg('PERPS_TRADING_PLAYBOOK_PATH', path.join(REPO_ROOT, 'docs', 'research', 'playbook-snapshot.md'));

const inferredBinDirs = [
  process.env.JUP_BIN_DIR || CONFIG.JUP_BIN_DIR,
  process.env.HELIUS_BIN_DIR || CONFIG.HELIUS_BIN_DIR,
  process.env.HOME ? path.join(process.env.HOME, '.hermes', 'node', 'bin') : '',
  process.env.HOME ? path.join(process.env.HOME, '.cargo', 'bin') : '',
].filter(Boolean);

export const PATH_ENV = inferredBinDirs.length ? `${inferredBinDirs.join(':')}:${process.env.PATH || ''}` : (process.env.PATH || '');
export const PERPS_BOT_CONFIG_FILE = CONFIG_FILE;
